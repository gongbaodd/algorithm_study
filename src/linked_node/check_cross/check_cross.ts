import { LNode } from "../lib";

export function findTailAndLen(head: LNode): [LNode, number] {
  let cur = head.next;
  let len = 1;

  while (cur?.next) {
    cur = cur.next;
    len += 1;
  }

  return [cur as LNode, len];
}

export function findCross(head1: LNode, head2: LNode): LNode | null {
  if (!head1.next || !head2.next) {
    return null;
  }

  const [tail1, len1] = findTailAndLen(head1);
  const [tail2, len2] = findTailAndLen(head2);

  if (tail1 !== tail2) {
    return null;
  }

  let cur1: LNode | null = head1;
  let cur2: LNode | null = head2;

  if (len1 > len2) {
    const sub = len1 - len2;
    for (let i = 0; i < sub; i += 1) {
      cur1 = cur1?.next ?? null;
    }
  } else {
    const sub = len2 - len1;
    for (let i = 0; i < sub; i += 1) {
      cur2 = cur2?.next ?? null;
    }
  }

  while (cur1 !== cur2) {
    cur1 = cur1?.next ?? null;
    cur2 = cur2?.next ?? null;
  }

  return cur1;
}
