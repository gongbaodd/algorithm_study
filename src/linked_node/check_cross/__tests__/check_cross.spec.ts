import { findCross, findTailAndLen } from "../check_cross";
import { init } from "../../lib";

describe("find cross", () => {
  describe("chain1 is longer than chain2", () => {
    it("should return the cross node", () => {
      const head1 = init([1, 2, 3, 4, 5, 6]);
      const head2 = init([1, 2, 3, 4]);
      const head3 = init([5, 6, 7]);

      findTailAndLen(head1)[0].next = head3;
      findTailAndLen(head2)[0].next = head3;

      expect(findCross(head1, head2)).toMatchSnapshot("1>2");
    });
  });

  describe("chain2 is longer than chain1", () => {
    it("should return the cross node", () => {
      const head1 = init([1, 2, 3]);
      const head2 = init([1, 2, 3, 4]);
      const head3 = init([5, 6, 7]);

      findTailAndLen(head1)[0].next = head3;
      findTailAndLen(head2)[0].next = head3;

      expect(findCross(head1, head2)).toMatchSnapshot("2>1");
    });
  });

  describe("no cross", () => {
    it("should return null", () => {
      const head1 = init([1, 2, 3]);
      const head2 = init([1, 2, 3, 4]);
      expect(findCross(head1, head2)).toBeNull();
    });
  });
});
