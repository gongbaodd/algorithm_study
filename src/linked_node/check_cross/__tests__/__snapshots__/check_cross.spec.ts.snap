// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`find cross chain1 is longer than chain2 should return the cross node: 1>2 1`] = `
LNode {
  "data": 0,
  "next": LNode {
    "data": 5,
    "next": LNode {
      "data": 6,
      "next": LNode {
        "data": 7,
        "next": null,
      },
    },
  },
}
`;

exports[`find cross chain2 is longer than chain1 should return the cross node: 2>1 1`] = `
LNode {
  "data": 0,
  "next": LNode {
    "data": 5,
    "next": LNode {
      "data": 6,
      "next": LNode {
        "data": 7,
        "next": null,
      },
    },
  },
}
`;
